File: packages/core/src/images.ts
import { ViktorInvalidRequestError } from "./errors.js";

export const MAX_IMAGES_PER_REQUEST = 10;
export const MAX_IMAGE_BYTES = 20 * 1024 * 1024;
export const ALLOWED_IMAGE_MIME_TYPES = ["image/jpeg", "image/png", "image/gif", "image/webp"] as const;
const MIME_ALIASES: Record<string, string> = { "image/jpg": "image/jpeg", "image/pjpeg": "image/jpeg", "image/x-png": "image/png" };

/**
 * Viktor silently skips images it cannot use (http URLs, unsupported types, more than 10 per
 * request). Failing on the client is the only way a developer notices, so validate before sending.
 */
export function validateImageUrl(url: string): void {
  if (url.startsWith("data:")) {
    const m = /^data:([^;,]+)[;,]/.exec(url);
    const mime = m ? (MIME_ALIASES[m[1]!.toLowerCase()] ?? m[1]!.toLowerCase()) : "";
    if (!(ALLOWED_IMAGE_MIME_TYPES as readonly string[]).includes(mime)) {
      throw new ViktorInvalidRequestError(
        `Viktor accepts ${ALLOWED_IMAGE_MIME_TYPES.join(", ")} images; got "${mime || "unknown"}".`,
      );
    }
    const base64 = url.slice(url.indexOf(",") + 1);
    if (base64.length * 0.75 > MAX_IMAGE_BYTES) {
      throw new ViktorInvalidRequestError("Image is larger than Viktor's 20 MiB limit.");
    }
    return;
  }
  if (url.startsWith("https://")) return;
  if (url.startsWith("http://")) {
    throw new ViktorInvalidRequestError("Viktor only fetches images over https. Use an https URL or a data URL.");
  }
  throw new ViktorInvalidRequestError("Image must be an https URL or a data URL.");
}

interface ChatMessageLike {
  content?: unknown;
}

/** Validate every `image_url` part in an OpenAI-shaped message list. */
export function validateChatImages(messages: readonly ChatMessageLike[]): void {
  let count = 0;
  for (const message of messages) {
    if (!Array.isArray(message.content)) continue;
    for (const part of message.content as Array<Record<string, unknown>>) {
      if (part?.type !== "image_url") continue;
      count += 1;
      const image = part.image_url as { url?: unknown } | string | undefined;
      const url = typeof image === "string" ? image : image?.url;
      if (typeof url !== "string") throw new ViktorInvalidRequestError("image_url part is missing a url.");
      validateImageUrl(url);
    }
  }
  if (count > MAX_IMAGES_PER_REQUEST) {
    throw new ViktorInvalidRequestError(
      `Viktor accepts at most ${MAX_IMAGES_PER_REQUEST} images per request; got ${count}. Extra images would be ignored.`,
    );
  }
}

File: python/core/src/viktor_integrations_core/images.py
"""Client-side image checks. Viktor silently skips images it cannot use, so fail early instead."""

from __future__ import annotations

import re
from collections.abc import Iterable, Mapping
from typing import Any

from .errors import ViktorInvalidRequestError

MAX_IMAGES_PER_REQUEST = 10
MAX_IMAGE_BYTES = 20 * 1024 * 1024
ALLOWED_IMAGE_MIME_TYPES = ("image/jpeg", "image/png", "image/gif", "image/webp")
_ALIASES = {"image/jpg": "image/jpeg", "image/pjpeg": "image/jpeg", "image/x-png": "image/png"}
_DATA_URL = re.compile(r"^data:([^;,]+)[;,]")


def validate_image_url(url: str) -> None:
    if url.startswith("data:"):
        m = _DATA_URL.match(url)
        mime = m.group(1).lower() if m else ""
        mime = _ALIASES.get(mime, mime)
        if mime not in ALLOWED_IMAGE_MIME_TYPES:
            allowed = ", ".join(ALLOWED_IMAGE_MIME_TYPES)
            raise ViktorInvalidRequestError(f'Viktor accepts {allowed} images; got "{mime or "unknown"}".')
        if len(url) - url.find(",") - 1 > MAX_IMAGE_BYTES * 4 / 3:
            raise ViktorInvalidRequestError("Image is larger than Viktor's 20 MiB limit.")
        return
    if url.startswith("https://"):
        return
    if url.startswith("http://"):
        raise ViktorInvalidRequestError("Viktor only fetches images over https. Use an https URL or a data URL.")
    raise ViktorInvalidRequestError("Image must be an https URL or a data URL.")


def validate_image_urls(urls: Iterable[str]) -> None:
    """Validate a flat list of image URLs (each URL, plus the 10-per-request limit)."""
    count = 0
    for url in urls:
        count += 1
        validate_image_url(url)
    if count > MAX_IMAGES_PER_REQUEST:
        raise ViktorInvalidRequestError(
            f"Viktor accepts at most {MAX_IMAGES_PER_REQUEST} images per request; got {count}. "
            "Extra images would be ignored."
        )


def validate_chat_images(messages: Iterable[Mapping[str, Any]]) -> None:
    """Validate image parts in an OpenAI-shaped message list.

    Understands Chat Completions parts (``{"type": "image_url", "image_url": {"url"}}``) and Responses API
    parts (``{"type": "input_image", "image_url": "..."}``).
    """
    count = 0
    for message in messages:
        content = message.get("content")
        if not isinstance(content, list):
            continue
        for part in content:
            if not isinstance(part, Mapping) or part.get("type") not in ("image_url", "input_image"):
                continue
            count += 1
            image = part.get("image_url")
            url = image if isinstance(image, str) else (image or {}).get("url")
            if not isinstance(url, str):
                raise ViktorInvalidRequestError("image_url part is missing a url.")
            validate_image_url(url)
    if count > MAX_IMAGES_PER_REQUEST:
        raise ViktorInvalidRequestError(
            f"Viktor accepts at most {MAX_IMAGES_PER_REQUEST} images per request; got {count}. "
            "Extra images would be ignored."
        )
